### ISSUE TITLE
Clang produces unhelpful diagnostic message

### ISSUE THREAD

**Issue by Endilll** [llvm/llvm-project#62823]
Consider the following [example](https://godbolt.org/z/xbqb3cosz) taken from [CWG1397](https://cplusplus.github.io/CWG/issues/1397.html):
```cpp
struct A {       
  void *p = A{};
  operator void*() const { return nullptr; }
};
```
ToT Clang issues the following (in C++14 mode and later):
```
<source>:2:15: error: default member initializer for 'p' needed within definition of enclosing class 'A' outside of member functions
  void *p = A{};
              ^
<source>:2:9: note: default member initializer declared here
  void *p = A{};
```
It describes what's happening, albeit in not too accessible phrasing IMO, but it fails to convey what's bad about it, leaving user without solid clue how to fix their code. I find [previous phrasing](https://github.com/llvm/llvm-project/commit/8dbc6b26171167b8ddf66a5f4b6d6fb9baf28336) of this diagnostic more clear for this particular case. 

To add to confusion, in C++11 mode an additional note is issued, and a slightly different line numbers are reported, as can be seen on [Compiler Explorer](https://godbolt.org/z/xbqb3cosz).

**Comment by shafik** [comment]
CC @AaronBallman

**Comment by AaronBallman** [comment]
That diagnostic was originally introduced in https://reviews.llvm.org/D5690
